libvast/include/vast/query_queue.hpp

```diff
@@ -0,0 +1,128 @@
+//    _   _____   __________
+//   | | / / _ | / __/_  __/     Visibility
+//   | |/ / __ |_\ \  / /          Across
+//   |___/_/ |_/___/ /_/       Space and Time
+//
+// SPDX-FileCopyrightText: (c) 2022 The VAST Contributors
+// SPDX-License-Identifier: BSD-3-Clause
+
+#pragma once
+
+#include "vast/fwd.hpp"
+
+#include "vast/query.hpp"
+#include "vast/system/actors.hpp"
+#include "vast/uuid.hpp"
+
+#include <vector>
+
+namespace vast {
+
+struct query_state {
+  /// The query expression.
+  vast::query query;
+
+  /// The query client.
+  system::receiver_actor<atom::done> client = {};
+
+  /// The number of partitions that need to be evaluated for this query.
+  uint32_t candidate_partitions = 0;
+
+  /// The number of partitions that have been reqested by the client.
+  uint32_t requested_partitions = 0;
+
+  /// The number of partitions that the query was sent to.
+  uint32_t scheduled_partitions = 0;
+
+  /// The number of partitions that are processed already.
+  uint32_t completed_partitions = 0;
+
+  template <class Inspector>
+  friend auto inspect(Inspector& f, query_state& x) {
+    return f(caf::meta::type_name("query_state"), x.query, x.client,
+             x.candidate_partitions, x.requested_partitions,
+             x.scheduled_partitions, x.completed_partitions);
+  }
+};
+
+class query_queue {
+public:
+  /// The entry type for the `partitions` lists. Maps a partition ID
+  /// to a list of query IDs.
+  struct entry {
+    uuid partition;
+    uint64_t priority = 0;
+    std::vector<uuid> queries;
+
+    friend auto operator<=>(const entry& lhs, const entry& rhs) {
+      return lhs.priority <=> rhs.priority;
+    }
+  };
+
+  // -- observers --------------------------------------------------------------
+
+  /// Calculates the number of partitions that need to be loaded to complete all
+  /// queries.
+  [[nodiscard]] size_t num_partitions() const;
+
+  /// Returns the number of currently queued queries.
+  [[nodiscard]] size_t num_queries() const;
+
+  /// Checks whether queries with outstanding work exist.
+  [[nodiscard]] bool has_work() const;
+
+  /// Checks whether the given query can be reached from the queue of
+  /// partitions. Should only be used for assertions.
+  [[nodiscard]] bool reachable(const uuid& qid) const;
+
+  /// Creates an ID for a query and makes sure to avoid collisions with other
+  /// existing query IDs.
+  [[nodiscard]] uuid create_query_id() const;
+
+  /// Retrieves a handle to the contained queries.
+  [[nodiscard]] const std::unordered_map<uuid, query_state>& queries() const;
+
+  /// Retrieves a handle to the contained queries.
+  [[nodiscard]] std::unordered_map<uuid, query_state>& queries();
```

Comment:
Afaict you only need the mutable overload in one place:

```c++
query_state.completed_partitions++;
```

I think the surrounding code section should be made a member function of the `query_queue`, then we can drop the unguarded access here.